### PR TITLE
Use `arc_swap` to implement `AtomicStr`

### PR DESCRIPTION
**Pull request by Kijewski**
The previous implementation allowed use-after-free in a multi-threaded context. This PR fixes the problem by using `ArcSwap`, which is implements thread-safe swapping of `Arc`s, and is widely used.

Resolves <https://github.com/longbridgeapp/rust-i18n/pull/71>.

### CODE REVIEW

**Comment by Kijewski** [comment]
```text
t                  [+31.689% +32.305% +32.878%]
t_with_locale      [+4.2409% +5.2036% +5.9512%]
t_with_threads     [+4.1089% +4.5351% +4.9449%]
t_lorem_ipsum      [+35.164% +35.690% +36.188%]
t_with_args        [+5.4206% +5.8389% +6.2721%]
t_with_args (str)  [+4.6291% +5.1848% +5.7272%]
t_with_args (many) [+3.0211% +3.3934% +3.7793%]
```

File: src/lib.rs [hunk 1]
```diff
@@ -16,8 +17,8 @@ pub fn set_locale(locale: &str) {
 }
 
 /// Get current locale
-pub fn locale() -> Arc<String> {
-    CURRENT_LOCALE.clone_string()
+pub fn locale() -> impl Deref<Target = str> {
+    CURRENT_LOCALE.as_str()
```
Comment (huacnlee) [inline src/lib.rs:21]: Do not change locale method for keep back forward compatibility.

File: src/lib.rs [hunk 2]
```diff
@@ -110,7 +111,7 @@ pub fn replace_patterns(input: &str, patterns: &[&str], values: &[String]) -> St
 macro_rules! t {
     // t!("foo")
     ($key:expr) => {
-        crate::_rust_i18n_translate(rust_i18n::locale().as_str(), $key)
+        crate::_rust_i18n_translate(&rust_i18n::locale(), $key)
```
Comment (huacnlee) [inline src/lib.rs:114]: Just use `CURRENT_LOCALE.as_str()` here, so we don't change the pub API.

**Review by huacnlee — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (Kijewski) [reply src/lib.rs:21]: `triomphe::Arc` is not the same as `std::sync::Arc`.

Comment (huacnlee) [reply src/lib.rs:21]: @varphone What do you think about this?

Last week, we just release v3 with breaking changes, I don't want to do it again.

Would we have any way to keep backward compatible?

Comment (Kijewski) [reply src/lib.rs:21]: It would always be possible to omit the second commit. It is only meant to negate the runtime impact of the first commit.

**Pull request merged**